Sources/NIO/ByteBuffer-extract.swift

```diff
@@ -0,0 +1,191 @@
+                    break
+                }
+                address = address.advanced(by: 1)
+            }
+            
+            let tokenLength = address - initialPointer
+            let ret = byteBuffer.getSlice(at: currentIndex, length: tokenLength)
+            // For the next time, with skipping the separator
+            currentIndex += tokenLength + 1
+            
+            return ret
+        }
+
+    }
+    
+    public init(byteBuffer: ByteBuffer, separator: UInt8, start: Int, length: Int) {
+        // This should never happens as we control when this is called. Adding an assert to ensure this.
+        assert(start <= byteBuffer.capacity - length)
+        
+        self.byteBuffer = byteBuffer
+        self.separator = separator
+        self.start = start
+        self.length = length
+        self.currentIndex = self.start
+    }
+
+    public init(byteBuffer: ByteBuffer,
+                separator: UInt8) {
+        
+        self.init(byteBuffer: byteBuffer, separator: separator,
+                  start: byteBuffer.readerIndex, length: byteBuffer.readableBytes)
+    }
+}
+
+extension ByteBuffer {
+    
+    public func sliceByTrimming(whiteSpaces: [UInt8], start: Int, length: Int) -> ByteBuffer? {
+        return withVeryUnsafeBytes { pointer -> ByteBuffer? in
+            assert(start <= self.capacity - length)
+            
+            let firstPtr = pointer.baseAddress!.assumingMemoryBound(to: UInt8.self)
+
+            var startPtr = firstPtr.advanced(by: start)
+            var endPtr   = startPtr.advanced(by: length - 1)
+            
+            // Advance startPtr until not a whiteSpace
+            while startPtr < endPtr && startPtr.pointee != 0 && whiteSpaces.contains(startPtr.pointee) {
+                startPtr = startPtr.advanced(by: 1)
+            }
+            
+            // Retreat endPtr until not a whiteSpace, null is considered a whiteSpace
+            while endPtr >= startPtr && (whiteSpaces.contains(endPtr.pointee) || endPtr.pointee == 0) {
+                endPtr = endPtr.advanced(by: -1)
+            }
+            return self.getSlice(at: startPtr - firstPtr, length: endPtr - startPtr + 1)
+        }
+    }
+    
+    public func sliceByTrimmingWhitespaces() -> ByteBuffer? {
+        return sliceByTrimming(whiteSpaces: [" ", "\t"].map({UInt8($0.utf8CString[0])}),
+                               start: 0, length: self.readableBytes)
+    }
+    
+    public func sliceByTrimmingWhitespaces(from start: Int = 0, length: Int) -> ByteBuffer? {
+        return sliceByTrimming(whiteSpaces: [" ", "\t"].map({UInt8($0.utf8CString[0])}),
+                               start: start, length: length)
+    }
+    
+    public func sliceByTrimmingWhitespaces(from start: Int) -> ByteBuffer? {
+        return sliceByTrimming(whiteSpaces: [" ", "\t"].map({UInt8($0.utf8CString[0])}),
+                               start: start, length: self.readableBytes - start)
+    }
```

Comment:
These three methods are not used internally, I don't think we need to add them at this time.

Reply:
I think they may be useful for some performance programmers (performance is a big reason for choosing Swift for web applications)

Reply:
As I noted elsewhere, I don't think we want to add APIs as "just in case" tools, especially when they're this simple to implement. This method is non-generic and boils down to calling a function with a set of specific arguments, which means there is no performance loss implementing it in any package that requires it.